File: frontend/src/game/engine/slide.ts
export function slideLineLeft(line: number[]): { line: number[]; gained: number } {
  const nonZero = line.filter((v) => v !== 0);
  const result: number[] = [];
  let gained = 0;
  let i = 0;
  while (i < nonZero.length) {
    if (i + 1 < nonZero.length && nonZero[i] === nonZero[i + 1]) {
      const merged = nonZero[i] * 2;
      result.push(merged);
      gained += merged;
      i += 2;
    } else {
      result.push(nonZero[i]);
      i += 1;
    }
  }
  while (result.length < 4) result.push(0);
  return { line: result, gained };
}

File: backend/internal/blockparty/engine/slide.go
package engine

// SlideLineLeft collapses a length-4 line toward index 0. Merge-once invariant:
// the source index advances by 2 on a merge, so a merged tile is never
// re-examined. Mirrors the TS slideLineLeft; keep integer-only for parity.
func SlideLineLeft(line []int) ([4]int, int) {
	nonZero := make([]int, 0, 4)
	for _, v := range line {
		if v != 0 {
			nonZero = append(nonZero, v)
		}
	}
	var result [4]int
	ri, gained, i := 0, 0, 0
	for i < len(nonZero) {
		if i+1 < len(nonZero) && nonZero[i] == nonZero[i+1] {
			merged := nonZero[i] * 2
			result[ri] = merged
			gained += merged
			ri++
			i += 2
		} else {
			result[ri] = nonZero[i]
			ri++
			i++
		}
	}
	return result, gained
}
